File: Roadmap/04 - CADENAS DE CARACTERES/python/qv1ko.py
def program(word1, word2):
    print(f"The word {word1} is a palindrome" if is_palindrome(word1) else f"The word {word1} is not a palindrome")
    print(f"The word {word2} is a palindrome" if is_palindrome(word2) else f"The word {word2} is not a palindrome")
    print(f"The words {word1} and {word2} are anagrams" if are_anagrams(word1, word2) else f"The words {word1} and {word2} are not anagrams")
    print(f"The word {word1} is an isogram" if is_isogram(word1) else f"The word {word1} is not an isogram")
    print(f"The word {word2} is an isogram" if is_isogram(word2) else f"The word {word2} is not an isogram")

def is_palindrome(word):
    length = len(word)
    for i in range(length // 2):
        if word[i] != word[length - i - 1]:
            return False
    return True

def are_anagrams(word1, word2):
    if len(word1) != len(word2):
        return False
    letters_word1 = sorted(word1)
    letters_word2 = sorted(word2)
    return letters_word1 == letters_word2

def is_isogram(word):
    letters = []
    for letter in word:
        if letter in letters:
            return False
        letters.append(letter)
    return True

program("solution", "loose")

File: Roadmap/04 - CADENAS DE CARACTERES/typescript/qv1ko.ts
program("facility", "disk");

function program(word1: string, word2: string): void {

    console.log(isPalindrome(word1) ? `The word ${word1} is a palindrome` : `The word ${word1} is not a palindrome`);
    console.log(isPalindrome(word2) ? `The word ${word2} is a palindrome` : `The word ${word2} is not a palindrome`);

    console.log(areAnagrams(word1, word2) ? `The words ${word1} and ${word2} are anagrams` : `The words ${word1} and ${word2} are not anagrams`);

    console.log(isIsogram(word1) ? `The word ${word1} is an isogram` : `The word ${word1} is not an isogram`);
    console.log(isIsogram(word2) ? `The word ${word2} is an isogram` : `The word ${word2} is not an isogram`);

}

function isPalindrome(word: string): boolean {

    const len = word.length;

    for (let i = 0; i < len / 2; i++) {
        if (word[i] !== word[len - i - 1]) {
            return false;
        }
    }

    return true;

}

function areAnagrams(word1: string, word2: string): boolean {

    if (word1.length !== word2.length) {
        return false;
    }

    const lettersWord1 = word1.split('').sort();
    const lettersWord2 = word2.split('').sort();

    return lettersWord1.join('') === lettersWord2.join('');

}

function isIsogram(word: string): boolean {

    const letters: string[] = [];

    for (const letter of word) {
        if (letters.includes(letter)) {
            return false;
        }
        letters.push(letter);
    }

    return true;

}
